### PR TITLE
Add node_exporter integration

### LINKED ISSUE #57: Create a YAML config section to control node exporter
# What we need to do

At the moment, `node_exporter` is controlled using a set of query parameters on the scrape URL. We'd like to avoid this and configure it using a similar structure to what we have today: YAML config.

# Implementation notes
- Add a new root key in the `Config` struct named `integrations` that holds a collection of exporters. e.g. A key for the collection in the case of `node_expoerter` would look like:

```YAML
server:
  log_level: info
  http_listen_port: 12345

prometheus:
  global:
    scrape_interval: 5s

integrations:
  node_exporter:
    an_example_option_here:
```

- Whatever we do, please keep in mind that in the future we'd like to load integrations from an individual file.
- Ensure we have proper documentation for the different keys too.

### LINKED ISSUE #84: Implement a node_exporter integration
Depends on #83 

A `node_exporter` integration should be added to the Agent, implementing the Integration interface.

### PR DESCRIPTION
**Pull request by rfratto**
This PR adds an embedded `node_exporter` to collect Unix system metrics. The embedded exporter is configured by YAML file. It aims to be near 100% compatible with the functionality of the full node_exporter, currently with the same defaults.

It currently uses a cheap hack that fills in kingpin flags that `node_exporter` reads, which is fine for the time being but may have to change if we embed another exporter that also happens to use kingpin for configuration. 

TODO: 
- [x] Add in missing flags present in node_exporter 1.0 
- [x] Documentation 
- [x] Crossplatform builds in `make dist` now that CGO is required for Darwin and FreeBSD

Closes #84 
Closes #57

### CODE REVIEW

**Comment by rfratto** [comment]
I'm not really happy with how this is configured. There's three things users would want to do when configuring:

1. Disable collectors that are enabled by default 
2. Enable collectors that are disabled by default 
3. Provide their own list of filters to add 
4. Configure extra options for collectors that accept options

The first three can be achieved by a list of `remove_collectors`, `add_collectors`, and `set_collectors` all that take a string array of collector slugs. The final list of collectors would then be calculated by `set_collectors` or `<default colllectors> + add_collectors - remove_collectors` 

The final can be done as it is now but flattening out the keys instead of nesting them inside `<slug>_collector`. 

For example:

```yaml
integrations:
  node_exporter:
    enable: true
    remove_collectors:
      - # remove collectors that are enabled by default 
    add_collectors:
      - # add collectors that are disabled by default

    # 
    # or, the user could set this 
    #
    set_collectors:
      - # set the list of collectors to run. takes precedence over remove_collectors/add_collectors 

    # extra options for collectors that support being configured
```

File: pkg/integrations/node_exporter/collectors.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
+package node_exporter // nolint:golint
+
+import (
+	"fmt"
+	"sort"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+) // CollectorState represents the default state of the collector, where it can
```
Comment (hoenn) [inline pkg/integrations/node_exporter/collectors.go:8]: Extra line here so the comment isn't on the `)` of import stmt

File: pkg/integrations/node_exporter/config.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+	TextfileDirectory string `yaml:"textfile_directory"` // --collector.textfile.directory
+}
+
+func (c *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	prefix = prefix + "node_exporter."
+
+	c.CommonConfig.RegisterFlagsWithPrefix(prefix, f)
+
+	f.BoolVar(&c.Enabled, prefix+"enabled", false, "enable the node_exporter integration collect metrics from the host Linux system")
```
Comment (hoenn) [inline pkg/integrations/node_exporter/config.go:82]: wording on the help text: "integration *to* collect"

File: pkg/integrations/agent/agent.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ type Config struct {
 
 func (c *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.BoolVar(&c.Enabled, prefix+"agent.enabled", false, "enable the Agent integration to scrape the Agent's own metrics")
+	c.CommonConfig.RegisterFlagsWithPrefix(prefix+"agent.", f)
```
Comment (rfratto) [inline pkg/integrations/agent/agent.go:25]: This was accidentally left out.

File: pkg/integrations/node_exporter/collectors.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		case CollectorStateDisabled:
+			flags = append(flags, "--no-"+flag)
+		}
+	}
+
+	sort.Strings(flags)
+	return
+}
+
+// DisabDisableUnavailableCollectors disables collectors that are not
```
Comment (robx) [inline pkg/integrations/node_exporter/collectors.go:177]: typo

File: pkg/integrations/node_exporter/collectors.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
+package node_exporter // nolint:golint
```
Comment (robx) [inline pkg/integrations/node_exporter/collectors.go:1]: just curious, what are these golint tags about?

File: pkg/integrations/node_exporter/node_exporter.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+func (i *Integration) Name() string { return "node_exporter" }
+
+// RegisterRoutes satisfies Integration.RegisterRoutes.
+func (i *Integration) RegisterRoutes(r *mux.Router) error {
+	handler, err := i.handler()
+	if err != nil {
+		return err
+	}
+
+	r.Handle("/metrics", handler)
```
Comment (gotjosh) [inline pkg/integrations/node_exporter/node_exporter.go:81]: Shouldn't this be using a different port or path for exposing the metrics? Or is there something else that's not part of this PR that will/currently takes care of that?

File: pkg/integrations/node_exporter/config.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+	f.StringVar(&c.TextfileDirectory, prefix+"textfile-directory",
+		"",
+		"directory for the textfile collector to read from")
+
+}
+
+// MapConfigToNodeExporterFlags takes in a node_exporter Config and converts
+// it to the set of flags that node_exporter usually expects when running as a
+// separate binary.
+func MapConfigToNodeExporterFlags(c *Config) []string {
```
Comment (gotjosh) [inline pkg/integrations/node_exporter/config.go:168]: Should we have a test for this function? A typo/change here end up in a breaking change.

File: pkg/integrations/node_exporter/node_exporter.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	return nil
+}
+
+func (i *Integration) handler() (http.Handler, error) {
```
Comment (gotjosh) [inline pkg/integrations/node_exporter/node_exporter.go:85]: How complex would it be to have a test for this handler?

File: example/docker-compose/docker-compose.no-agent.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
+version: "2"
```
[the remark this answers is not in the document]
Comment (rfratto) [reply example/docker-compose/docker-compose.no-agent.yaml:1]: Good catch, I checked this in by mistake. I was using it to launch a cortex/grafana so I could iterate on the integration locally. I'll take it out :+1:

File: pkg/integrations/node_exporter/config.go [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+	EnableCollectors flagext.StringSlice `yaml:"enable_collectors"`
+
+	// Collectors to mark as disabled
+	DisableCollectors flagext.StringSlice `yaml:"disable_collectors"`
+
+	// Overrides the default set of enabled collectors with the collectors
+	// listed.
+	SetCollectors flagext.StringSlice `yaml:"set_collectors"`
+
+	// Collector-specific config options
```
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/integrations/node_exporter/config.go:36]: Yeah, I know :slightly_frowning_face: I had to pull them manually by running `node_exporter --help`. They're not documented in the README.

File: pkg/integrations/node_exporter/node_exporter.go (hunk 6 above)
Comment (rfratto) [reply pkg/integrations/node_exporter/node_exporter.go:81]: All integrations get a subrouter, so `/metrics` here expands to `/integrations/node_exporter/mterics`. I'll add a comment to make this more obvious :+1:

File: pkg/integrations/node_exporter/collectors.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	CollectorUname        = "uname"
+	CollectorVMStat       = "vmstat"
+	CollectorWiFi         = "wifi"
+	CollectorXFS          = "xfs"
+	CollectorZFS          = "zfs"
+)
+
+// Collectors holds a map of known collector names to their default
+// state.
+var Collectors = map[string]CollectorState{
```
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/integrations/node_exporter/collectors.go:91]: great idea, I'll create an issue for it after this gets merged

File: pkg/integrations/node_exporter/collectors.go (hunk 5 above)
Comment (rfratto) [reply pkg/integrations/node_exporter/collectors.go:1]: `golint` complains about the package having an underscore in the name, but I find that having an underscore is appropriate here

**Comment by rfratto** [comment]
> I'm a bit curious whether there's a plan for keeping that huge flag list in collectors in sync with upstream node exporter?

Yep, this integration will remain 1:1 with the full list of configuration options provided by upstream node_exporter. I don't anticipate the list changing too often, though.

**Comment by rfratto** [comment]
@gotjosh I think testing this is at the unit going to be a little difficult and hard to extract value from. I think the best thing we can do is an integration test similar to:

1. Provide a full config with everything enabled 
2. Create the exporter 
3. Run the scrape once

Provided it doesn't return an error, that should validate that the flags were parsed correctly. WDYT?

**Comment by rfratto** [comment]
I've done more local testing and I'm confident this is good to merge.

File: pkg/integrations/node_exporter/collectors.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/integrations/node_exporter/collectors.go:91]: Opened #120 for this.
